### DISCUSSION TITLE
Call logging class

### DISCUSSION THREAD

**Discussion by so195** [eclipse-langium/langium#1555] [Q&A] [answered]
I am trying to add a logger to my langium project. I implemented the following class which works when I call it from another file in the console.

```
import { Logger, createLogger, transports } from "winston";

export default class AppLogger {
        public static Logger: Logger;
      
        public static InitialiseLogger( ) {
          const logger = createLogger({
            transports: [
              new transports.File({ filename: "log.log" }),
            ],
         });
      
          AppLogger.Logger = logger;

          AppLogger.LogInfo("AppLogger", `Info 1`);
         }

       public static LogInfo(label: string, message: string) {
          AppLogger.Logger.info({ label, message });
        }
  }
```
When I call my logger function from my main.ts file and just start my program, it doesn't work.
This is the code I wrote in my main file:

```
AppLogger.InitialiseLogger();
AppLogger.LogInfo("main file", `Info 3`);
```
I also tried logging from other points in my program, but there also nothing happens.
I don't get any error. I am assuming I am not calling my function correctly.

What is the correct way and the correct place to call such a function, so that I can use it throughout my whole application?

**Comment by msujew** [comment]
Hey @so195,

Can you provide a reproducible example? Note that language servers are not allowed to write to the normal stdio console (they get re-routed instead), as the stdio channel is usually used by the language client to transfer messages.

**Reply by so195** [reply]
Is there any way to allow writing to the console?  But if that isn't possible it would be even more important to be able to write a log file.

To minimize any other influences, I used the hello world project from the langium tutorial.
I only added my file `appLogger.ts` in the src folder:

```
import { Logger, createLogger, transports } from "winston";

export default class AppLogger {
  public static Logger: Logger;

  public static InitialiseLogger() {
    const logger = createLogger({
      transports: [new transports.File({ filename: "log.log" })],
    });

    AppLogger.Logger = logger;

    AppLogger.LogInfo("AppLogger", `Info 1`);
  }
  public static LogInfo(label: string, message: string) {
    AppLogger.Logger.info({ label, message });
  }
}

```

And I added the last two lines to my `main.ts` file to initialize and test my logger:

```
import { startLanguageServer } from "langium/lsp";
import { NodeFileSystem } from "langium/node";
import {
  createConnection,
  ProposedFeatures,
} from "vscode-languageserver/node.js";
import { createHelloWorldServices } from "./hello-world-module.js";
import AppLogger from "../appLogger.js";

// Create a connection to the client
const connection = createConnection(ProposedFeatures.all);

// Inject the shared services and language-specific services
const { shared } = createHelloWorldServices({ connection, ...NodeFileSystem });

// Start the language server with the shared services
startLanguageServer(shared);

AppLogger.InitialiseLogger();
AppLogger.LogInfo("main file", `Info 3`);

```

**Reply by msujew** [reply, marked as answer]
Thanks for the code! However, I cannot reproduce your issue, as the library correctly logs the data to a file (in the open VS Code workspace, since it's the `cwd` of the vscode extension).

```json
{"label":"AppLogger","level":"info","message":"Info 1"}
{"label":"main file","level":"info","message":"Info 3"}
```

You might want to be a bit more specific about the filepath you're using. Maybe something like `__dirname + '/log.log'` instead to ensure that the log is created in the same directory where the language server is located?

> Is there any way to allow writing to the console?

Note that you can still write to `console.log`, but the log is rerouted to a different channel - when running in vscode, it will instead be printed in the output log of your language server:

![image](https://github.com/eclipse-langium/langium/assets/4377073/41fed766-8448-4ec5-97c6-abf6ccf6d814)

**Reply by so195** [reply]
Thank your very much, that fixed my Problem!
